pkg/services/screenshot/screenshot.go

```diff
@@ -84,70 +31,28 @@ type ScreenshotService interface {
 	Take(ctx context.Context, opts ScreenshotOptions) (*Screenshot, error)
 }
 
-// CachableScreenshotService caches screenshots.
-type CachableScreenshotService struct {
-	cache       *gocache.Cache
-	service     ScreenshotService
-	cacheHits   prometheus.Counter
-	cacheMisses prometheus.Counter
-}
-
-func NewCachableScreenshotService(r prometheus.Registerer, expiration time.Duration, service ScreenshotService) ScreenshotService {
-	return &CachableScreenshotService{
-		cache:   gocache.New(expiration, time.Minute),
-		service: service,
-		cacheHits: promauto.With(r).NewCounter(prometheus.CounterOpts{
-			Name:      "cache_hits_total",
-			Namespace: namespace,
-			Subsystem: subsystem,
-		}),
-		cacheMisses: promauto.With(r).NewCounter(prometheus.CounterOpts{
-			Name:      "cache_misses_total",
-			Namespace: namespace,
-			Subsystem: subsystem,
-		}),
-	}
-}
-
-// Take returns the screenshot from the cache or asks the service to take a
-// new screenshot and cache it before returning it.
-func (s *CachableScreenshotService) Take(ctx context.Context, opts ScreenshotOptions) (*Screenshot, error) {
-	k := fmt.Sprintf("%s-%d-%s", opts.DashboardUID, opts.PanelID, opts.Theme)
+// ManagedScreenshotService is a managed screenshot service that features caching,
+// single-flight, rate-limiting and instrumentation. It is the recommended service
+// to use unless screenshots are disabled, in which case either NoOpScreenshotService
+// or UnavilableScreenshotService are recommended.
+type ManagedScreenshotService struct {
+	cache        CacheService
+	capture      CaptureService
+	singleFlight SingleFlight
+	tokens       TokenBucket
 
-	if v, ok := s.cache.Get(k); ok {
-		defer s.cacheHits.Inc()
-		return v.(*Screenshot), nil
-	}
-
-	defer s.cacheMisses.Inc()
-	screenshot, err := s.service.Take(ctx, opts)
-	if err != nil {
-		return nil, err
-	}
-
-	s.cache.Set(k, screenshot, 0)
-
-	return screenshot, nil
-}
-
-// NoopScreenshotService is a service that takes no-op screenshots.
-type NoopScreenshotService struct{}
-
-func (s *NoopScreenshotService) Take(_ context.Context, _ ScreenshotOptions) (*Screenshot, error) {
-	return &Screenshot{}, nil
-}
-
-// ObservableScreenshotService is a service that records metrics about screenshots.
-type ObservableScreenshotService struct {
-	service   ScreenshotService
 	duration  prometheus.Histogram
 	failures  *prometheus.CounterVec
 	successes prometheus.Counter
 }
 
-func NewObservableScreenshotService(r prometheus.Registerer, service ScreenshotService) ScreenshotService {
-	return &ObservableScreenshotService{
-		service: service,
+func NewManagedScreenshotService(cs CaptureService, sf SingleFlight, tb TokenBucket, ch CacheService, r prometheus.Registerer) ScreenshotService {
+	return &ManagedScreenshotService{
+		cache:        ch,
```

Comment:
Do you want to support a "no cache" service? There's a RAM cache and a mock right now. Do we just reduce the expiration time to zero to get "don't cache" functionality? That works, but it allocates a lot of objects before immediately expiring them.